### ISSUE TITLE
Add support for forwarding implementation

### ISSUE THREAD

**Issue by andrewlock** [khellang/Scrutor#57]
Basically, to work around the limitations in this: https://github.com/aspnet/DependencyInjection/issues/360.

As far as I can tell, there's no way to use Scrutor assembly scanning and also forward multiple interface types to a single concrete service? So for example, with Scrutor code:

```csharp
public class Bar : IBar, IFoo
{
} 
```

```csharp
services.Scan(scan =>
                scan.FromAssemblyOf<Bar>()
                    .AddClasses(c => c.AssignableTo<IBar>())
                    .AsImplementedInterfaces()
                    .AsSelf()
                    .WithSingletonLifetime());

var sp = services.BuildServiceProvider();
var i1 = sp.GetService<Bar>();
var i2 = sp.GetService<IBar>();
var i3 = sp.GetService<IFoo>();

Assert.Same(i1, i2); // Fails
Assert.Same(i2, i3); // Fails
```

With a "raw" implementation, you can work around this issue by forwarding the request to the implementation type. It's inefficient, but works:

```csharp
services.AddSingleton<Bar();
services.AddSingleton<IFoo>(x=>x.GetService<Bar>());
services.AddSingleton<IBar>(x=>x.GetService<Bar>());

var sp = services.BuildServiceProvider();
var i1 = sp.GetService<Bar>();
var i2 = sp.GetService<IBar>();
var i3 = sp.GetService<IFoo>();

Assert.Same(i1, i2); // Passes
Assert.Same(i2, i3); // Passes
```

It would be nice to be able to do the same with Scrutor, e.g.

```csharp
services.Scan(scan =>
                scan.FromAssemblyOf<Bar>()
                    .AddClasses(c => c.AssignableTo<IBar>())
                    .AsImplementedInterfaces(forwardToImplementationtype: bool) // optional parameter, or could be a new method
                    .AsSelf()
                    .WithSingletonLifetime());
```

I think it would be safe to add either as an overload to `AsImplementedInterfaces`, or as a new method entirely. And it's a real blocker I have at the moment 🙁 

What do you think? I'm happy to send a PR for it if it sounds good to you?

**Comment by khellang** [comment]
> As far as I can tell, there's no way to use Scrutor assembly scanning and also forward multiple interface types to a single concrete service?

Right, that code would make three separate registrations and would result in three "singletons"; one for `Bar`, one for `IBar` and one for `IFoo`.

If I'm reading this right, you'd need both `AsSelf` and `AsImplementedInterfaces` in order to properly resolve the service?

I think it'd have to be a separate method, but I'm having a really hard time coming up with a good name proposal 😝

Anyway, if you're willing to do the work, I'm sure we can come up with a good method name 👍

**Comment by khellang** [comment]
FWIW, it's been brought up before (with some sample code); https://github.com/khellang/Scrutor/issues/47

**Comment by andrewlock** [comment]
> If I'm reading this right, you'd need both AsSelf and AsImplementedInterfaces in order to properly resolve the service?

Yes it would

> I think it'd have to be a separate method, but I'm having a really hard time coming up with a good name proposal 😝

Haha, why do you think I suggested an optional parameter 😉 

But I agree, an extra method makes sense, especially as the `AsSelf()` call is required to make it work. I'll have a think of some names (`AsImplementedInterfacesForwardedToClass`? mmm.... so wordy....) and put together a PR 🙂 

> FWIW, it's been brought up before (with some sample code); #47

Thanks - I thought I'd seen it before but my search skills were failing me!
